### ISSUE TITLE
Large tables are slow to show up in datasets dashboard and at dataset page

### ISSUE THREAD

**Issue by andy-esch** [CartoDB/cartoframes#210]
When writing a large table through cartoframes (e.g., 1e6 rows), the dataset page (https://team.carto.com/u/eschbacher/dataset/{tablename}) gives a 404 even though the table is accessible through the SQL API. It can take 2-5 minutes for the table to show up in the dashboard and through it's own page. It could be related to the table needing to registered with the carto metadata. Once visiting the dataset dashboard, the tables seem to appear more quickly. We need a programmatic way to ensure the table's dataset view shows up as soon as cartoframes points to the table page. Does CDB_TouchMetadata do this?

cc @javitonino

**Comment by javitonino** [comment]
Yes, this is surely caused by creating the table via SQL API. We have a process, called ghost tables, that registers those kind of tables. This process is executed (asynchronously) when visiting the dashboard. The only way is to visit the dashboard, wait for a few seconds and refresh. There is currently no way to manually trigger this via API or similar.

Why this only happens for 1e6 rows, is because at 500k, you switch from import API (which properly registers the table) to SQL API. There is no option to disable overviews creation at Import API.

I can currently see no way out of this without adding some parameters/endpoints to CARTO, waiting for overviews to create (which may be too slow for CF use case) or some ugly hacks (if the dataset has some polygon in the first rows, overview creation does not tigger).

@juanignaciosl any suggestion?

**Comment by juanignaciosl** [comment]
@javitonino has described the process perfectly. There's no way you can ensure that the dataset view shows up as soon as it's created, because of the asynchronous nature of ghost tables and the big size of the table.

The only current workaround that I can come up with is, once you've created the table with SQL API, make a request to the dashboard from cartoframes (so ghost tables is triggered) and then poll the (private) dashboard API (something similar to https://team.carto.com/u/juanignaciosl/api/v1/viz/?tag_name=&q=&page=1&type=&exclude_shared=false&per_page=20&tags=&shared=no&locked=false&only_liked=false&order=updated_at&types=table&deepInsights=false&api_key=xxxxxx) until it shows up.

**Comment by andy-esch** [comment]
Thanks for the thorough answers -- it helps a lot. Does running `CDB_CartodbfyTable(...)` register it so that it will appear listed through that end point? What I'm seeing is the temp files which are imported through the Import API and then deleted, but not the table created from them with a query like this:

```SQL
CREATE TABLE foo AS 
SELECT ...columns... FROM foo_temp_0
UNION ALL
SELECT ...columns... FROM foo_temp1;
```

`foo_temp_0` and `foo_temp_1` appear through that endpoint but `foo` doesn't. I looked at the network traffic in the browser and tried various params but could not replicate what happens magically when the dataset dashboard is loaded. 

This is what I'm trying out:

```python
def _ping_dashboard(self):
    """Ping CARTO dashboard to get ghost tables to show"""
    params = {
        'api_key': self.api_key,
        'deepInsights': 'false',
        'exclude_shared': 'false',
        'locked': 'false',
        'only_liked': 'false',
        'order': 'updated_at',
        'page': '1',
        'per_page': '20',
        'shared': 'no',
        'types': 'table'
    }
    try:
        resp = self.auth_client.send(
                '{baseurl}api/v1/viz/'.format(baseurl=self.base_url),
                http_method='GET',
                params=params
            )
        resp.raise_for_status()
        self._debug_print(ping_response=resp.text)
    except requests.exceptions.HTTPError as err:
        warn(err)
```

**Comment by juanignaciosl** [comment]
> Does running CDB_CartodbfyTable(...) register it so that it will appear listed through that end point?

I'm afraid not. Running `CDB_CartodbfyTable` doesn't trigger a request to Rails or the metadata database, the one used for the listings.

`(...)/api/v1/viz(...)` is not the dashboard request but the visualizations endpoint. ~In order to trigger ghost tables you should trigger a request to `https://<username>.carto.com/dashboard` (and follow redirections, because you ...~. 

I just realized that you can't trigger ghost tables programatically right now :(. Ghost tables are triggered when you go to the dashboard (https://team.carto.com/u/juanignaciosl/dashboard). Nevertheless, that request won't work unless you have the session cookie. Using the api_key is not allowed.

We could easily provide an API endpoint that triggers ghost tables, but this is becoming hacky and it doesn't make much sense. As you're switching from Import to SQL API to avoid overviews creation, maybe you could use Import API only, and if the overview is actually a problem we might add a parameter at Import API to disable creation. What do you think, @javitonino ?

**Comment by andy-esch** [comment]
I'm using SQL API + Import API to enable making tables as large as the user has in the DataFrame (so mainly row limits on the Import API are the barrier). But I did choose a threshold one smaller than the overview generation limit (499999) since overviews caused some slow downs. Disabling overview creation would be pretty handy because it'd allow larger batches, but it's not at all a blocker here.

I could do a refactor of my code and rely on the first table sent up with the Import API and then insert all other tables into that first one, avoiding creating a table with the SQL API.

**Comment by juanignaciosl** [comment]
That would be great, because Import API registers the table.

**Comment by andy-esch** [comment]
👍 I'll refactor and let you know how it goes.

Is there a reliable way to drop datasets without them being reverse-ghosts like in the screenshot above?

**Comment by juanignaciosl** [comment]
> Is there a reliable way to drop datasets without them being reverse-ghosts like in the screenshot above?

I overlooked this in your previous comment, sorry! No with a public API. You can send a `DELETE` to `http://username.carto.com/api/v1/viz/TABLE_NAME?api_key=xxxx`, though.

**Comment by andy-esch** [comment]
Solved in #211.
The general method is to create a placeholder table with the Import API, create the final table with SQL API, delete the original table with SQL API, rename final table with SQL API.

Most of the code is here:
https://github.com/CartoDB/cartoframes/blob/e6290773bbf93a3fdbae1275ef8ca03bab0a7f5f/cartoframes/context.py#L330-L338
